lib/ukcpuid/cpuid.c

```diff
@@ -0,0 +1,182 @@
+#include <stdio.h>
+#include <string.h>
+#include <stdlib.h>
+#include <stdbool.h>
+
+#include <uk/errptr.h>
+#include <uk/cpuid.h>
+
+#include "../plat/common/include/x86/cpu_defs.h"
+
+static struct uk_cpuid_info info;
+static struct uk_cpuid_cache cache;
+
+static unsigned int __highest_basic_func_value;
+static unsigned int __highest_extended_func_value;
+static unsigned int last_eax_function = -1;
+static unsigned int last_ecx_function = -1;
+
+
+unsigned char is_flag_enabled(unsigned int eax_function, unsigned int ecx_function,
+							  unsigned int flag, unsigned int *reg)
+{
+
+	if (last_eax_function != eax_function ||
+		last_ecx_function != ecx_function) {
+		last_eax_function = info.EAX = eax_function;
+		last_ecx_function = info.ECX = ecx_function;
+
+		cpuid_macro(info.EAX, info.EBX, info.ECX, info.EDX, info.EAX, info.ECX);
+	}
+
+	return (*reg) & flag;
+}
+
+void init_cpuid(void)
```

Comment:
Do you think we could get the library working without init? This would simplify the usage at early boot code (e.g., in assembler).

Reply:
Yes, we could move the init behavior in every method exposed by the interface (get maximum value for eax and compare with current eax). Of course, after first request of getting maximum eax, we could store the value and use it for further checkings. However, since Pentium processor the maximum value for eax is at least 0x03, therefore for small value we can skip the checking. What do you say?

Reply:
@samoilescusebastian , I don't think adding the init behavior in every method is a viable option. It would incur overhead each time it's called.

@skuenzer , could you elaborate on some concrete case where the library would be called in early boot code? You want to have an "on demand" inquiry at boot time, without adding the `init_cpuid()` call overhead?